apps/web/src/api/hooks/use-vercel-integration.ts

```diff
@@ -0,0 +1,58 @@
+import axios from 'axios';
+import { useContext, useState, useCallback, useEffect } from 'react';
+import { useMutation } from 'react-query';
+import { useSearchParams } from 'react-router-dom';
+import { AuthContext } from '../../store/authContext';
+import { errorMessage } from '../../utils/notifications';
+import { vercelIntegrationSetup } from '../vercel-integration';
+
+export function useVercelIntegration() {
```

Comment:
I totally understand your point,
But if we remove the `useEffect` and went with the `useCallback` way, the issue will be tracking the axios authorisation header,

I had tried it before, If we invoke `startVercelSetup` from the pages, after login/signup, then at that moment the axios auth header doesn't have the JWT token stored, and sometimes it leads to a race condition where the vercel mutate API has been invoked but the axios auth header setup was not completed, and hence we get 401 unauthorized API error,

Also, the below code would just return as Axios header setup was in-process and was not finished by the time , the below code was executed, and it would just return, and it wont retry/re-fire after axios setup is completed.
```
if (!canStartSetup || !code) {
      return;
    }
```
so instead of manually tracking the Axios auth header state on each pages, I added it to the hook itself and the `useEffect` helps to trigger the mutate API call once the axios setup is finished.

Reply:
> When the user is logged in the token is applied to the state on the initial render
```js
const [token, setToken] = useState<string | null>(() => {
      const localStorageToken = getToken();
      applyToken(localStorageToken);
  
      return localStorageToken;
    });
```
Actually, this hook mounts on the login page itself, even when the user is not logged in, so in that case, the initial render will be fired and the lazy state initialization of the `useState` will have the `null` value, and upon login it will lead to a side-effect trigger, it wont be an initial render anymore.

I tried this approach now, but it didn't worked, the axios setup was still not completed by the time vercel setup was invoked.
```js
  const startVercelSetup = useCallback(() => {
    console.log(' start vercel setup called');

    if (!canStartSetup || !code) {
      return;
    }
    mutate(code);
  }, [canStartSetup, code]);
```
it didn't invoked the `mutate` as `isAxiosAuthorized` was still undefined

Reply:
Hey, I've spent some time debugging the case here with "not logged in user" and found the issue that was causing it to not work.

The Login Page `useEffect:22` is calling on the successful login through GH provider, it takes the token from the params and sets it in the context, but the `setToken` is the state setter function, it doesn't update the axios headers. 
That's why when the `useEffect:28` is triggering `startVercelSetup` in the result it is not calling mutation, the header is not set at that moment.

To fix that issue we need to update the `useAuthController` -> `setToken` to be a callback like:
```
 const setTokenCallback = useCallback(
    (newToken: string | null) => {
      setToken(newToken);
      applyToken(newToken);
    },
    [setToken]
  );
```

But that's not all the problems, because children's effects are called before the parent, meaning that the `useAuthController` effects will be called after the login page effects, and at that time of `useAuthController` doesn't have the updated state yet, so it's "second" `useEffect:65` will be called with `null` token and will clear the header again. But we know in that hook that the token now will be set using the `setTokenCallback` function, so we can remove the `applyToken` from there. So the `useEffect:65` will look like this now:
```
useEffect(() => {
    if (token) {
      queryClient.removeQueries({
        predicate: (query) =>
          query.queryKey !== '/v1/users/me' &&
          query.queryKey !== '/v1/environments' &&
          query.queryKey !== '/v1/organizations',
      });
      const payload = jwtDecode<IJwtPayload>(token);
      setJwtPayload(payload);
    }
  }, [token]);
```

---
So the diff for the all changes looks like this:

![Screenshot 2022-10-08 at 01 25 21](https://user-images.githubusercontent.com/2607232/194674958-b6a19b3b-c84d-4e0e-904f-807da3f00b95.png)

![Screenshot 2022-10-08 at 00 24 47](https://user-images.githubusercontent.com/2607232/194670789-23c16849-51f1-4769-91b9-670ee7af5b2c.png)

---
Of course, I've verified the logged-in/logged-out cases locally and it works for me ;) 

Logged in:

https://user-images.githubusercontent.com/2607232/194675773-7773d702-523f-42ac-8492-13b9c8289891.mov

Logged out:


https://user-images.githubusercontent.com/2607232/194675790-182d497f-55b9-46d5-907c-ee4ef8a71f76.mov

Reply:
@LetItRock Thank you for helping me refactor the Auth flow. 👏